contracts/raffle/src/lib.rs

```diff
@@ -128,6 +128,10 @@ impl RaffleContract {
             return Err(Error::NotInitialized);
         }
 
+        if storage.get::<_, bool>(&DataKey::AlreadyPlayed).unwrap() {
+            return Err(Error::AlreadyPlayed);
+        }
+
```

Comment:
Nitpick:

Exactly this checking is also repeated in other methods. I consider that code a bit ugly and difficult to read. An alternative could be wrap it in a function like:

```rust
fn must_not_be_already_played() -> Result<()> {
       if storage.get::<_, bool>(&DataKey::AlreadyPlayed).unwrap() {
            return Err(Error::AlreadyPlayed);
        }
}
```
Then we could reuse that function:
```rust
fn my_contract_function_1() -> Result<(), Error>{
// ...
must_not_be_already_played()?;
//...
}
fn my_contract_function_2(){
// ...
must_not_be_already_played().unwrap();
//...
}
```